core/src/main/java/io/grpc/ChannelImpl.java

```diff
@@ -302,8 +336,7 @@ public void transportTerminated() {
           log.warning("transportTerminated called without previous transportShutdown");
           activeTransport = null;
         }
-        // TODO(notcarl): replace this with something more meaningful
-        transportShutdown(Status.UNKNOWN);
+        clearActiveTransport();
```

Comment:
One thing: I like the idea of being able to reason about listeners.  So far, the rule I have observed is:

Listener methods are always called at most once, unless specified otherwise.  Having one callback call another callback just in case kind of breaks that idea.  It isn't wrong, but it slows down understanding.  I moved the contents of clearActiveTransport into transport shutdown.